cpp/arcticdb/storage/mongo/mongo_mock_client.hpp

```diff
@@ -0,0 +1,92 @@
+/* Copyright 2023 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#pragma once
+
+#include <arcticdb/storage/storage_utils.hpp>
+#include <arcticdb/storage/mongo/mongo_client_wrapper.hpp>
+#include <arcticdb/storage/storage_mock_client.hpp>
+#include <mongocxx/exception/operation_exception.hpp>
+
+namespace arcticdb::storage::mongo {
+
+struct MongoFailure {
+    // holds std::monostate if it is a no acknowledgement failure because in case of an acknowledgement from server
+    // the mongo apis don't throw an exception
+    std::variant<mongocxx::operation_exception, std::monostate> failure;
+
+    [[nodiscard]] bool is_no_ack_failure() const {
+        return std::holds_alternative<std::monostate>(failure);
+    }
+
+    mongocxx::operation_exception get_exception() const {
+        return std::get<mongocxx::operation_exception>(failure);
+    }
+
+};
+
+// the key is composed of the database name, collection name and the variant key
+using mongo_document_key = std::tuple<std::string, std::string, std::string>;
```

Comment:
What do you think about using a struct instead?

Reply:
Fixed in [ea60ab1](https://github.com/man-group/ArcticDB/pull/1395/commits/ea60ab11b0de9987e1146fa249374be97efd0934). This is actually a very good idea. I have now defined a `MongoDocumentKey` which saves the entire `VariantKey` and defines `id_string()` which will contain the failure trigger(if set using `get_failure_trigger()`). This overcomes the previous [limitation in `list_keys()`](https://github.com/man-group/ArcticDB/pull/1395/files/7896c47e4be34fe966fd6bae28fb27629a525f97#diff-c25a56cd300b6b66f5ec51696594f1e745d58755fe77b64d2e4c15bafc6d03e7R179), we can now return any key instead of just `TABLE_DATA`.